File: src/tusk/nodes/effects/set.py
from tusk.node import Node
from tusk.token import Token
from tusk.nodes.expressions import ExpressionNode
from tusk.nodes.base.name import NameNode
from tusk.variable import Variable, is_ordinal_number

class SetNode(Node):
    def __init__(self, token: Token):
        self.token = token
        self.interpreter = token.interpreter

    async def create(self):
        name = self.interpreter.next_token()
        if is_ordinal_number(name):
            print("guh")
            n = is_ordinal_number(name)-1
            self.interpreter.expect_token("KEYWORD:item|KEYWORD:character")
            self.interpreter.expect_token("LOGIC:in")
            e = NameNode(self.interpreter.next_token())
            self.interpreter.expect_token("KEYWORD:to")
            e.value[n] = (await ExpressionNode(self.interpreter.next_token()).create()).value
        else:
            """
                        vname = name.value
                        to_set = self.interpreter.data["vars"]
                        while self.interpreter.get_next_token().type == "PROPERTY":
                            to_set = to_set[vname]
                            self.interpreter.next_token()
                            vname = self.interpreter.expect_token("IDENTIFIER").value
                            to_set = to_set.properties
            """
            n = (await NameNode(self.interpreter.current_token).create())

            self.interpreter.expect_token("KEYWORD:to")
            value = (await ExpressionNode(self.interpreter.next_token()).create()).value
            if isinstance(value, Variable):
                value.name = n.name
                n.location[n.name] = value
                self.interpreter.data["vars"]["it"] = value
                self.interpreter.data["vars"]["this"] = value
            else:
                n.location[n.name] = Variable(n.name,value)
                self.interpreter.data["vars"]["it"] = value
                self.interpreter.data["vars"]["this"] = value
                

File: src/java/tusan/nodes/effects/Set.java
package net.tutla.tums.tusan.nodes.effects;

import net.tutla.tums.tusan.Node;
import net.tutla.tums.tusan.Utils;
import net.tutla.tums.tusan.Variable;
import net.tutla.tums.tusan.lexer.Token;
import net.tutla.tums.tusan.lexer.TokenType;
import net.tutla.tums.tusan.nodes.base.Name;
import net.tutla.tums.tusan.nodes.expression.Expression;

import java.util.List;
import java.util.Map;

public class Set extends Node {
    private Utils utils = new Utils();
    public Set(Token token){
        super(token);
    }

    public Set create(){
        Token name = interpreter.nextToken();
        Object num = utils.isOrdinal(name);
        if (num != null){ // TODO: this code is fucked so like ill fix it later
            int n = ((Double) num).intValue();
            n--;
            interpreter.expectTokenClassic("KEYWORD:item|KEYWORD:character");
            interpreter.expectToken(TokenType.LOGIC,"in");
            Object val;
            if (interpreter.getNextToken().type == TokenType.IDENTIFIER){
                Name e = new Name(interpreter.nextToken()).create();
                val = Name.name;
            } else {
                Object e = new Expression(interpreter.nextToken());
                val = e;
            }
            interpreter.expectToken(TokenType.KEYWORD,"to");

            Object exprValue = new Expression(interpreter.nextToken()).create().value;

            if (val instanceof List) {
                ((List<Object>) val).set(n, exprValue);
            } else if (val instanceof Map) {
                ((Map<Object, Object>) val).put(n, exprValue);
            }
        } else {
            Name n = new Name(name).create();
            interpreter.expectToken(TokenType.KEYWORD,"to");
            Object val = new Expression(interpreter.nextToken()).create().value;
            if (val instanceof Variable){
                ((Variable) val).name = n.name;
                n.location.put(n.name, val);
                interpreter.data.vars.put("it",val);
                interpreter.data.vars.put("this", val);
            } else {
                n.location.put(n.name, new Variable(n.name, val, null));
                interpreter.data.vars.put("it",val);
                interpreter.data.vars.put("this", val);
            }
        }
        return this;
    }
}
